### PR TITLE
Sorter with classes

### PR DESCRIPTION
**Pull request by samuelgarcia**
This PR is the port towards classes approach.

This very first comit is the proof of concept of class approach + one axample with **klusta** only.

@colehurwitz31 @alejoe91 @magland @mhhennig : could you have a look  at least for the API.

I write a SorterBsae with virtual methods and try to implement it on klusta.

**pararlel** and **by_property** is not done at the moment. So it is not too difficult to read.

This is totally untested, just a proof of concept for the API.

Could you give me some feedback ? At least next Wednesday on meeting.

I will next:
  *  make the same for tridesclous and
  * make sorterlist.py that deal with sorter list (unstalled or not installed
  * make several ngine to run sorters in paralelle with thread + multiprocessing in base class.

best

Samuel

### CODE REVIEW

**Comment by alejoe91** [comment]
I was able to look at the entire files and I really like the organization :)

I can help in porting the spikeSortByProperty stuff. Let's port the other spike sorters in the next days.

**Comment by samuelgarcia** [comment]
If you like the main idea I can port other sorters very soon.

File: spiketoolkit/sorters/basesorter.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+"""
+
+import time
+import copy
+from pathlib import Path
+
+class BaseSorter:
+    
+    sortername = '' # convinience for reporting
```
Comment (colehurwitz31) [inline spiketoolkit/sorters/basesorter.py]: sortername --> sorter_name

Comment (colehurwitz31) [reply spiketoolkit/sorters/basesorter.py]: Just want to be consistent with variable names (this is small)

File: spiketoolkit/sorters/basesorter.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        assert self.installed, """This sorter {} is not installed.
+        Please install it with:  \n{} """.format(self.sortername, self.installation_mesg)
+        
+        if output_folder is None:
+            output_folder = sorter_name
+        
+        self.output_folder = Path(sorter_name)
+        self.recording = recording
+        self.debug = debug
+        self.by_property = by_property
```
Comment (colehurwitz31) [inline spiketoolkit/sorters/basesorter.py]: by_property --> sorting_property

File: spiketoolkit/sorters/basesorter.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        self.parallel = parallel
```
Comment (colehurwitz31) [inline spiketoolkit/sorters/basesorter.py]: parallel --> is_parallel?

File: spiketoolkit/sorters/basesorter.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        
+    def _setup_recording(self):
+        # need subclass
+        raise(NotImplemenetdError)
+
+    def _run(self):
+        # need subclass
+        raise(NotImplemenetdError)
+    
+    def get_result(self):
```
Comment (colehurwitz31) [inline spiketoolkit/sorters/basesorter.py]: get_result should check to see if output file exists or if the spike sorter has been run

File: spiketoolkit/sorters/basesorter.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        sorting = self.SortingExtractor_Class(self.output_folder)
+        return
+    
+    
+    # new idea
+    def get_params_for_particular_recording(self, rec_name):
```
Comment (colehurwitz31) [inline spiketoolkit/sorters/basesorter.py]: What is this for?

File: spiketoolkit/sorters/basesorter.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+       """
+       this is speculative an nee to be discussed
+       """
+       return {}
+
+
+# generic laucnher via function approach
+def run_sorter_engine(SorterClass, recording, output_folder=None,
```
Comment (colehurwitz31) [inline spiketoolkit/sorters/basesorter.py]: This looks good, but I would call it:

run_sorter_engine --> run_sorter

File: spiketoolkit/sorters/klusta/klusta.py [hunk 7]
```diff
@@ -38,127 +28,113 @@ def klusta(
...
+    def __init__(self, recording=None, output_folder=None,
+                                    by_property=None, parallel=False):
+        BaseSorter.__init__(self, recording=recording, output_folder=output_folder,
+                                    by_property=by_property, parallel=parallel)
+
+    def set_params(self, **params):
+        self.params = params
+        
+    
+    def _setup_recording(self):
```
Comment (colehurwitz31) [inline spiketoolkit/sorters/klusta/klusta.py]: I don't understand what _setup_recording does to be honest, 

- Why does it save a probe file and also save out to a binary format?

File: spiketoolkit/sorters/sorterlist.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (colehurwitz31) [inline spiketoolkit/sorters/sorterlist.py]: Is this file just keeping track of all sorters that are installed?

File: spiketoolkit/sorters/tridesclous/tridesclous.py [hunk 9]
```diff
@@ -1,95 +1,125 @@
...
+        initial_catalogue = self.tdc_dataio.load_catalogue(chan_grp=0)
+        peeler = tdc.Peeler(self.tdc_dataio)
+        peeler.change_params(catalogue=initial_catalogue,
+                            use_sparse_template=use_sparse_template,
+                            sparse_threshold_mad=1.5,
+                            use_opencl_with_sparse=use_opencl_with_sparse,)
+        peeler.run(duration=None, progressbar=True)
+
+
+def run_tridesclous(
```
Comment (colehurwitz31) [inline spiketoolkit/sorters/tridesclous/tridesclous.py]: So this is a different function then run_sorter_engine?

Comment (colehurwitz31) [reply spiketoolkit/sorters/tridesclous/tridesclous.py]: Should they be the same?

**Comment by colehurwitz31** [comment]
Looks great! I think it should be how we move forward with spike sorting algorithms

I did a little code review and had some questions about naming and functions. It seems like the run_sorter function for trideclous and the basesorter are different?

File: spiketoolkit/sorters/tridesclous/tridesclous.py (hunk 9 above)
Comment (samuelgarcia) [reply spiketoolkit/sorters/tridesclous/tridesclous.py]: At this moment it is the same.
I was wondering between a unique generic one or one runner function for each SorerClass.

The run_sorter_engine have also in the signautre the SorterClass itself.

the run_tridesclous have the same signeture as beefore (and so for run_klusta)

File: spiketoolkit/sorters/sorterlist.py (hunk 8 above)
Comment (samuelgarcia) [reply spiketoolkit/sorters/sorterlist.py]: It is a list for all sorter and all installed sorter.

So for testing sorters on a particular dataset the list will generate automatiocally with future helper fuction.

File: spiketoolkit/sorters/klusta/klusta.py (hunk 7 above)
Comment (samuelgarcia) [reply spiketoolkit/sorters/klusta/klusta.py]: It is split in 2 steps:
  * setup recording (basically copy/paste files), it can take long time for big files
  * and the comptation itself.

If we want to make benchmark on the second step only it will be usefull.

File: spiketoolkit/sorters/basesorter.py (hunk 6 above)
Comment (samuelgarcia) [reply spiketoolkit/sorters/basesorter.py]: OK

File: spiketoolkit/sorters/basesorter.py (hunk 5 above)
Comment (samuelgarcia) [reply spiketoolkit/sorters/basesorter.py]: In my mind, some sorter will need to adapt the default parameters depending on the dataset caracteristics. Sorters with same parameters don't behave the same depending if the dataset is a tetrode or dense array. in vitro or in vivo, etc...
So for a fair comparison, it should be possible to change parameters for a given dataset label.
It is just a futur proposal.

File: spiketoolkit/sorters/basesorter.py (hunk 4 above)
Comment (samuelgarcia) [reply spiketoolkit/sorters/basesorter.py]: Yep.
Good idea.

File: spiketoolkit/sorters/basesorter.py (hunk 3 above)
Comment (samuelgarcia) [reply spiketoolkit/sorters/basesorter.py]: yes of course. a typo.

File: spiketoolkit/sorters/basesorter.py (hunk 5 above)
Comment (colehurwitz31) [reply spiketoolkit/sorters/basesorter.py]: I see. Maybe a function like, "adapt_parameters(self, recording)" which takes in a recording extractor and allows the algorithm to adapt its default parameters. We don't want to pass in the recording_file_name because we are trying to abstract away the file type for that.

File: spiketoolkit/sorters/klusta/klusta.py (hunk 7 above)
Comment (colehurwitz31) [reply spiketoolkit/sorters/klusta/klusta.py]: So set_up recording does what exactly? I thought that the recording extractor was supposed to eliminate the need for copying over big files. Am I missing something?

File: spiketoolkit/sorters/sorterlist.py (hunk 8 above)
Comment (colehurwitz31) [reply spiketoolkit/sorters/sorterlist.py]: I see. Sounds good.

File: spiketoolkit/sorters/tridesclous/tridesclous.py (hunk 9 above)
Comment (colehurwitz31) [reply spiketoolkit/sorters/tridesclous/tridesclous.py]: I think each algorithm should have the same function, "run_sorter". The name and signature for this function should be the same across spike sorters. Is that possible? It would make running all sorters the same which is great for standardized analysis.

File: spiketoolkit/sorters/basesorter.py (hunk 1 above)
Comment (samuelgarcia) [reply spiketoolkit/sorters/basesorter.py]: OK

**Comment by samuelgarcia** [comment]
Yes idealy run_sorter can run all SorterClass but by compatibility I also did run_tridesclous and run_klusta that have same signature as tridesclou() and klusta().

So all examples only need to change klusta(..) to run_klusta(...)

**Comment by colehurwitz31** [comment]
I would rather change all the examples to run_sorter. We don't have many examples to fix.

**Comment by samuelgarcia** [comment]
I also prefer.

**Comment by samuelgarcia** [comment]
I am trying to move the moutainsort4 function to class.

There is something different for this one : there is no extractor in spikeextractor.

The ml_ms4alg directly render a sorting with its own input ouput function.
This could limit a bit the new class approach that decompose (for multiprocessing reason) in 3 steps: 
  * setup recording
  * run
  * get spikeextractor


What is the reason of having no extractor in spikeextractor of this sorter ?

**Comment by samuelgarcia** [comment]
Other problem in moutainsort4, there is no output_folder.

This will make the framework difficult because there is no explicit persistency on disk.

In my mind, for multiprocessing, this should be mandatory to all sorter.

**Comment by colehurwitz31** [comment]
I am not sure what you mean, in the mountainsort algorithm, a recording extractor is used throughout (each recording extractor is called a recording). Also, it uses sorting extractors (each sorting extractor is called a sorting).

We can look into adding an output_folder to the sorter. Maybe Jeremy can weigh in on that. @magland

File: spiketoolkit/sorters/klusta/klusta.py (hunk 7 above)
Comment (alejoe91) [reply spiketoolkit/sorters/klusta/klusta.py]: hei @colehurwitz31 
many of the spikesorters require to save the data in a specific format (e.g. dat) and or provide probe files and params files. 

The _setup_recording() is a nice way to organize these steps.

Thanks @samuelgarcia

**Comment by alejoe91** [comment]
@samuelgarcia and @colehurwitz31 
I can comment on the mountainsort approach, as I'm also implement the same idea in SpyICA (https://github.com/alejoe91/spyica).

As we have the SpikeInterface framework implemented, the idea is to bypass the spike-sorting specific formatting etc, so that the spike sorting functions themselves directly work on the recordingExtractor object and return a sortingExtractor one. 

@magland It would be nice to have ms4 alg to accept an output_folder argument instead of using the /tmp/ folder by default. I have made some changes to ml_ms4alg and can make a PR. 

@samuelgarcia for now I would put an dummy `output_folder` arg for standardizing the inputs.
A possible solution could be having a second Base class, `BaseSISorter` (where SI is SpikeInterface) for sorters which are developed within the SpikeInterface framework.

What do you guys think?

**Comment by colehurwitz31** [comment]
I think the plan should be for all sorters to take in a recording extractor and output a sorting extractor (can also save a normal output folder). 

I think the BaseSorter class should be sufficient for this purpose rather than defining two classes.

**Comment by alejoe91** [comment]
@colehurwitz31 I think you are missing the point. We are talking about adhering to the _setup_recording(), run(), get_sorting() implementation.

For mountainsort, this is not possible because the ms4 function from ml_ms4alg package already takes a recordingExtractor and returns a sortingExtractor. So the _setup_recording and get_sorting functions are not needed

**Comment by colehurwitz31** [comment]
Wouldn't it be:

1) _setup_recording() --> Do nothing (or whatever setup is needed to use ml_ms4alg)
2) run()  --> stores the sorting extractor internally after running ml_ms4alg
3) get_sorting() --> returns the internally stored sorting extractor

I don't see the issue with this. Every algorithm will do something similar and if they don't use one of the steps, that is ok.

**Comment by samuelgarcia** [comment]
Another keypoint iwith this new approach is the ability to serialize a sorter (so its parameters and output_forlder) to be executed elsewhere (subprocess or other machine).
So ideally, even if it is not the case now. The python interpreter could stop and restart in bewteen theses steps to facilitate batch across process and then collect results.

In short, a sorter should not have to resident in memory for further analyses.

**Comment by samuelgarcia** [comment]
@alejoe91 : I would be more in favor of having one unique class.
Another class BaseSISorter would make the framework more fuzzy.


My feeling is that sorting tools that already have a "SpikeInterface" API builtin,  should **also** have a spikeextractor class even if it is a trivial  function to function mapping.

So the get_result would be more generic.

We should really have in mind that comparing several sorter on real big fat datasets (>256 channel over minuts and more) will never fit in memory even the results.

**Comment by alejoe91** [comment]
Maybe we could extend the NumpyExtractor so that it could load map numpy files?

Than we could just dump the unit_ids, spiketrains, and possibly properties and features and load them from file?
It would be nice to have the same for recording extractors too.

**Comment by alejoe91** [comment]
BTW, I can help implementing some sorting classes. Should I tackle kilosort?

**Comment by colehurwitz31** [comment]
> Maybe we could extend the NumpyExtractor so that it could load map numpy files?
> 
> Than we could just dump the unit_ids, spiketrains, and possibly properties and features and load them from file?
> It would be nice to have the same for recording extractors too.

File: spiketoolkit/sorters/ironclust/ironclust.py [hunk 10]
```diff
@@ -1,7 +1,163 @@
...
+        'detect_threshold': 5,  # Threshold for detection
+        'merge_thresh': .98,  # Cluster merging threhold 0..1
+        'freq_min': 300,  # Lower frequency limit for band-pass filter
+        'freq_max': 6000,  # Upper frequency limit for band-pass filter
+        'pc_per_chan': 3,  # Number of pc per channel
+        'parallel': True
+    }
+
+    installation_mesg = """
+       >>> pip install mountainlab_pytools
```
Comment (alejoe91) [inline spiketoolkit/sorters/ironclust/ironclust.py]: can you add: `kbucket` to the pip install? That is also needed for ironclust

**Comment by samuelgarcia** [comment]
@alejoe91 : I have push some minut ago teh kilosort class.
But I will need a lot of help for testing theses code.

I don't have matlab installed on my mahcine so kilosrt and ironclust is not tested at all!!!!!
I did not run then.
What you can do is make a branch from my repo with:
 **git remote add samuelgarcia https://github.com/samuelgarcia/spiketoolkit.git**
Then, debug ironclust an kilosrt.
And then push back via a PR on my repo.

I am doing the circus.

File: spiketoolkit/sorters/ironclust/ironclust.py (hunk 10 above)
Comment (samuelgarcia) [reply spiketoolkit/sorters/ironclust/ironclust.py]: with pip install kbucket ?

**Comment by alejoe91** [comment]
Just saw it :)

Ok I will update the run_all_sorters notebook and make sure that everything runs. Will also add tridesclous

**Comment by samuelgarcia** [comment]
At the time right now only tridesclous is working ...
klusta should also work but I have "Exception: Klusta did not run successfully"
due to bad execution.

You can have a look in spiketoolkit/sorters/test and try to run then one by one.

**Comment by alejoe91** [comment]
@samuelgarcia 
How is the installation_mesg supposed to work? can we make it self in the constructor?

klusta, ironclust, and kilosort run. Just wanted to figur out this before making a PR

**Comment by alejoe91** [comment]
Regarding my previous comment: I see it's static already! Haven't used static variables before in Python and had to google it...

**Comment by alejoe91** [comment]
@samuelgarcia 
made a PR for klusta, kilosort, and ironclust fixes

**Comment by colehurwitz31** [comment]
Hi all, 

Is this ready to be merged? If not, is there anything I can do to help?

Best,

Cole

**Comment by samuelgarcia** [comment]
No.
This is not ready.

We need to:
  * make the parralel and by properties thing.
  * debug classes
  * progagate changes everywhere (doc, examples)
  * make the moutain sort class working.
  * make travis working

I will try at the end of the weekd and will need help from Alessio from the by property.

At the moment, to help you can clone this and test some classe randomly.

Look at sorters/tests/test_*.py

**Comment by colehurwitz31** [comment]
So just to clarify, I would clone and then try running each of the tests by installing spike sorters? Then, I should report the results to you somehow?

**Comment by samuelgarcia** [comment]
You can even directly debug it on your report with a git remote add and make a PR on my report before the "big" merge into main report.

**Comment by samuelgarcia** [comment]
I dig a bit in moutainsort4 code.

inside ml_ms4alg module moutainsort4 function is this one:

```python
  MS4=MountainSort4()
  MS4.setRecording(recording)
  geom=_get_geom_from_recording(recording)
  MS4.setGeom(geom)
  MS4.setSortingOpts(
    clip_size=clip_size,
    adjacency_radius=adjacency_radius,
    detect_sign=detect_sign,
    detect_interval=detect_interval,
    detect_threshold=detect_threshold
  )
  tmpdir = tempfile.mkdtemp()
  MS4.setNumWorkers(num_workers)
  print('Using tmpdir: '+tmpdir)
  MS4.setTemporaryDirectory(tmpdir)
  try:
    MS4.sort()
  except:
    print('Cleaning tmpdir:: '+tmpdir)
    shutil.rmtree(tmpdir)
    raise
  print('Cleaning tmpdir::::: '+tmpdir)
  shutil.rmtree(tmpdir)
  times,labels,channels=MS4.eventTimesLabelsChannels()
  output=se.NumpySortingExtractor()
  output.setTimesLabels(times=times,labels=labels)
```

I propose to base the class Mountainsort4Sorter on that piece of code instead of having this output_input fixed function.
So it could fit more the new class approach that decompose in several steps and fix the output_folder thing. @magland @alejoe91 : is it OK ?

I do proposal soon.

**Comment by samuelgarcia** [comment]
There is still one problem with moutain sort is that there is no extractor directly in spikeextractor.
I think we should write one based on mountainlab_pytools.
@magland : what do you think ?

**Comment by samuelgarcia** [comment]
I am about to make the _spikeSortByProperty.

I think I understand the thing.
To be clear. If property is not None then a list of SubRecordingExtractor is done.
And so a list of sorting is computing in parrlell via thread or not.

I plan to implement this as follow in BaseSorter.
recording will always be a list recordings event there is only one.
Internaly get_result**s(..)** will a be a list of SortingExtractor and get_result() will make a MultiSortingExtractor in any case.

My my concern with that approach is that there are several way to make several "groups".
This actual one with "property" and the second ihhen one with the probe file.
I think that the 2 approach should converge, in short "probe_file" should be a common keyargs even if the sorter do not deal with it and this probe file should be used to create a "group_id" property for channels taht would be reflected in the recodring so the SubRecordingExtractor would also work inthat case.

What do you think ?

**Comment by alejoe91** [comment]
Everything looks good but I thin that the 2 apporaches to spikesort by group are already convergent. 
When you load the probe file (or pass it as an arg - which we could make a common keyarg) the group information (if present in the probe file) is loaded in the recordingExctractor as channel property. 
Then one can call the spike sorting function with propert=group. This should automatically:
- split the recording extractor in several subrecordings based on the property (done with se.getSubExtractorsByProperty)
- run the spike sorting in parallel
- reassemble the results and adding the right property to the unit (e.g. unit 23 is in group 6)

this approach allows to deal with the probe_files (in that case the property would be group in most cases), but it also keeps it flexible in case the user wants to split the sorting by other means.

Is it clear?

**Comment by samuelgarcia** [comment]
OK. This was the piece I totally miss.
Good I anderstand everything. I will implement this soon so.

**Comment by samuelgarcia** [comment]
Is it OK if I change the "by_property" to "grouping_property" ?

**Comment by samuelgarcia** [comment]
Is it OK to add this behavior:

* when there is an explicit probe_file with several group AND grouping_property is None then the
  grouping_property is automatically set to "group" ?

**Comment by samuelgarcia** [comment]
There is a conflict.
It is more complicated than I though.

For sorter based on PRB file (tridesclous, circus, klusta, ...) the input trace (recording) have all the channel. Inside the prb file each group is mapped with the absolut channel map.
If we use the "group" property for theses sorters, then traces are reduced to a smaller subset of channel count and then the PRB is wrong because it play with absolut channel index.

In short, sorter with no PRB file support the grouping is done **externally** by the SubRecordingExtractor but for sorter than support PRB file the grouping is done internaly by the sorter.

So we need that if probe_file is not None then "by_property" (or "grouping_property") must be None.

Am I clear ?

**Comment by alejoe91** [comment]
Hi Sam,

I don't think I understood the issue here, but the way I use it now the probe is loaded externally and the spike sorter is run. 
I think this is a good way to do it: load the probe file before running sorting:
- load extractor
- load probe file
- run spike sorting

**Comment by samuelgarcia** [comment]
@alejoe91 
I code the proof of concept of multi group handling in BaseSorter.

I change a bit the API.
Now the setup, run and result is done explicitly on one rescording (or subrecording).
So:
  *  **def _setup_recording(self, recording, output_folder):**
  *  **def _run(self, recording, output_folder):**
  * **def _get_one_result(self, recording, output_folder):**

With this it is easy to loop over subrecording for multi groups and launch them in parallel.

I push changes only for tridesclous.

If you are happy with this I will make change for all sorters.

**Comment by samuelgarcia** [comment]
Also note that I start to make a SorterCommonTestSuite class that will launch a serie of test for each sorter class with a very simple hineritence system.

So basic feature of sorter will tested for each one with pytest when installed.

**Comment by samuelgarcia** [comment]
3 task to do before merge:
  * make the moutain sort class working (I need help here)
  * make travis working
  * debug. Only tridesclous pass test on my machine.

**Comment by samuelgarcia** [comment]
Yeah!!!!
Travis is passing test!!!!
🥂

Situation of CI for sorter:
  * klusta : tested and pass on travis
  * tridesclous :  tested and pass on travis
  * circus: hard to install on travis because the MPI
  * kilosort : never will run travis because matlab
  * ironsort : never will run travis because matlab
  * ms4 : no implemented yet but should ru easly on travis.

@colehurwitz31 @alejoe91 : could make a full pytest on you mahcine if you have all sorters installed ?

@magland @alejoe91 : still need help for MS4

**Comment by colehurwitz31** [comment]
Wow!! Nice work :) I think @alejoe91 managed to install all the spike sorters (with some dark magic) so many he can quickly make a full pytest.

**Comment by samuelgarcia** [comment]
> with some dark magic
+ a matlab licence :(

**Comment by alejoe91** [comment]
Wow @samuelgarcia this is amazing!

I made a PR with a working implementation ( a bit hacky) of MS4 in this framework.

One think that we should add is the ability to provide only a few arguments. If a parameter is not given then the default is used. Right now ALL the parameters need to be passed. Maybe a check_parameters in the base class that loop through the keys of _default_params and kwargs and uses the default one if not ofund in kwargs?

**Comment by alejoe91** [comment]
> * circus: hard to install on travis because the MPI

I think mpi can be easily installed as follows from the .travis before the install:

```
sudo: enabled
addons:
  apt:
    packages:
    - libopenmpi-dev
    - openmpi-bin
```

I can try to test if spyking-circus will run with this plus pip install spyking-circus

**Comment by samuelgarcia** [comment]
OK I will try circus on travis.

The fact is that circus is buggy on my clean install because in some situation circus try to open a GUI that close immediatly. And it create problem... I guess on github it will be worse.

Have you test circus locally ?

**Comment by samuelgarcia** [comment]
For the parameters, Iwanted but I forgot to make a set_params at BaseSorter level that take all default_parms and update with given one. It was my plan but I forgot it.

Ithink it will fix what you say for params.

**Comment by samuelgarcia** [comment]
Parameters fixed. Thanks.

**Comment by magland** [comment]
Hi all. I just wanted to give brief feedback. It seems like you are doing
very good work. I guess my intuition tells me to keep it simple with
function calls and parameters. But I see the advantage of putting all
sorters in the same framework with base class. One could always create a
simple interface on top if needed.

**Comment by samuelgarcia** [comment]
Hi @magland .
Thank you.

The previous functional interface is  kept with but **run\_kilo(..)** instead of **kilo(..)**, **run\_klusta(...)**. instead of **klusta(..)**, ...

Have a look in sorterlist.py, it is quite simple.

**Comment by magland** [comment]
Nice!

**Comment by alejoe91** [comment]
@samuelgarcia I'm running all tests and will push a commit on the PR on your fork as soon as they all run (also on travis)

**Comment by samuelgarcia** [comment]
Travis pass test but it fact it run only a few of then. (only tdc and klusta)
circus and ms4 shoudl also run there.
Could someone could clear the cache of travis ?

**Comment by samuelgarcia** [comment]
I need some help.

I am fighting with pytest on travis.

The travis version of pytest is 3.3.0. this version is too old and do not deal well with class unittest inheritence.
So, no test are launch except a few one.
The script do pip install pytest and so the version should 4.3.0
With this version tests behave better.

I tried a python -m pytest but it do not change anything.


Any ideas ?

**Comment by alejoe91** [comment]
@samuelgarcia 
Is the problem on travis or locally?

**Comment by alejoe91** [comment]
you can try pip install pytest==4.3

**Comment by samuelgarcia** [comment]
locally I have virtualenv so I am sur that the pytest is the good one.

On travis, my guess is taht pytest is already installed at system level but it is an old one.
So th "pip install pytest" which install the latest version have no effect.

On my machine, I am able to run tests and crash then :)
On travis everything is skipped because of the design I choose (class inheritance  with SorterCommonTestSuite)

**Comment by alejoe91** [comment]
I see. Have you tried pip install pytest==4.3 from the .travis?

**Comment by alejoe91** [comment]
@samuelgarcia the travis stuff is fixed in the new PR i made to your fork :)

**Comment by samuelgarcia** [comment]
I will be off at the begining of the next week.

If you are happy with I think we can merge this PR. [if travis pass all test]

I will be back on launcher.py soon with a new PR.
The actual on is just a proof of concept.

**Comment by alejoe91** [comment]
What do you guys think to have a dev branch with all these major API changes and merge when everything is done and stable? @samuelgarcia @colehurwitz31 @magland

**Comment by samuelgarcia** [comment]
My opinion in a fresh and new project like this one is to merge everything and soon to master to avoid several branches.
You can put tags sometimes instead if you want retro compatibility.

Figuring which feature is in the dev and wich is in the master. is really time consuming.

You should consider the master (as the dev) + branch in private repo for feature + tags(to release). No more.

**Comment by alejoe91** [comment]
Then let's wait to merge as we use extensively spiketoolkit in our analysis pipeline. Have to update and test the other codes too :) Give me a couple of days

**Comment by colehurwitz31** [comment]
How exactly the class system can help with dependencies between spike sorters? I remember we discussed this, but I am still unclear.

Do you think it would be a good idea to make docker images for each spike sorter? That could be a way to handle all the dependency issues.

**Comment by alejoe91** [comment]
@samuelgarcia 

made a PR to you fork with some changes. After that we can merge here :)

**Comment by samuelgarcia** [comment]
@alejoe91 : I made some comment in the sub PR.

**Comment by samuelgarcia** [comment]
@colehurwitz31 : at the moment there are no dependency "between" sorter.

For each sorter class there is a ```installed``` attribute that can be check to see if the sorter if usablme or not.
See sorterlist.py.
The ```installed``` mainly depend for each sorter of a global variable that is initialized at module level with help of ```try .. except ModuleNotFoundError```

For instance in tridesclous:
```
try:
    import tridesclous as tdc
    HAVE_TDC = True
except ModuleNotFoundError:
    HAVE_TDC = False

class TridesclousSorter(BaseSorter):
   ...
    installed = HAVE_TDC
    ...
```

In basesorter if installed=False then an error is raised an the ```installation_mesg``` is show to the user.


For a total automatic installation of each sorter we could also add a classattribute like
 ```
class TridesclousSorter(BaseSorter):
   ...
    install_cmd = """
pip install https://github.com/tridesclous/tridesclous/archive/master.zip
""""
 ```

At the moment the command is in the ```installation_mesg```
